aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1414,6 +1414,16 @@ static void addmm_impl_cpu_(
   AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND(kBFloat16,
       result.scalar_type(), "addmm_impl_cpu_",
       [&]{
+        #if defined(__aarch64__) && AT_MKLDNN_ACL_ENABLED()
```

Comment:
the conditions here are insufficient, in other places where `mkldnn_matmul` is called it's guarded by `use_mkldnn_bf16_matmul`

Reply:
We guarded the `mkldnn_matmul` with bf16 checks because the fp32 counterpart is still not faster than MKL on x86. For ARM, is fp32 also faster?

Reply:
Does `mkldnn_matmul` support double and complex types? I doubt it, I think there's a check in it that limits it float and bfloat16, and here it's called without any additional checks from a dispatch function that dispatches to all dtypes.

Reply:
> Does mkldnn_matmul support double and complex types?

No. Agree that a check on float and bfloat16 should be added at least.

Reply:
Thanks @ngimel and @jgong5 for your review.

> We guarded the mkldnn_matmul with bf16 checks because the fp32 counterpart is still not faster than MKL on x86. For ARM, is fp32 also faster?

Yes, on AArch64 fp32 is also faster. I have made a change such that we check whether type is `fp32` and only do `mkldnn_matmul` for that case. On AArch64 we didn't implement at the moment `matmul` primitive that operates on `bf16` datatype. To use BF16 instructions you can set `ONEDNN_DEFAULT_FPMATH_MODE` to `BF16` (see [here](https://oneapi-src.github.io/oneDNN/dev_guide_attributes_fpmath_mode.html#a-note-on-default-floating-point-math-mode)) to do internal maths with BF16 instructions instead of FP32 instructions (operands are implicitly casted from FP32 to BF16). Then we have observed that for models such as BERT_pytorch runs 2.3x faster with this flag set to BF16.

Reply:
@milpuz01 if you want to expose internal bf16 compute for fp32 inputs without setting env var you can check the value of `set_float32_matmul_precision` switch https://pytorch.org/docs/master/generated/torch.set_float32_matmul_precision.html#torch.set_float32_matmul_precision and dispatch to bf16 compute if "medium" is set.

Reply:
@ngimel Good to know there is a flag to turn on bf16 compute for fp32 matmul. Is there a similar one for convolution? oneDNN supports bf16 compute for both fp32 matmul and conv.

Reply:
@jgong5 to be clear, currently changing the value of this flag won't do anything for cpu, it's up to you to actually do this behavior change, but at least pytorch has API to expose it. 
We didn't introduce similar flag for convolutions, sorry.

Reply:
@ngimel @milpuz01 FYI: https://github.com/pytorch/pytorch/pull/91848